core/auth/jwt.go

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (C) 2019 The "MysteriumNetwork/node" Authors.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package auth
+
+import (
+	"math/rand"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+	"github.com/pkg/errors"
+)
+
+// JwtAuthenticator contains JWT handling methods
+type JwtAuthenticator struct {
+	storage       Storage
+	encryptionKey []byte
+}
+
+// JWTToken contains token details
+type JWTToken struct {
+	Token          string
+	ExpirationTime time.Time
+}
+
+// JWTCookieName name of the cookie JWT token is stored in
+const JWTCookieName string = "token"
+
+type jwtClaims struct {
+	Username string `json:"username"`
+	jwt.StandardClaims
+}
+
+type jwtKey []byte
+
+const expiresIn = 48 * time.Hour
+const encryptionKeyBucket = "jwt"
+const encryptionKeyName = "jwt-encryption-key"
+
+// NewJWTAuthenticator creates a new JWT authentication instance
+func NewJWTAuthenticator(storage Storage) *JwtAuthenticator {
```

Comment:
I'd redesign this component a bit. You're injecting storage just to store the key - why not actually fetch(or generate and store if it does not exist) the key and pass it as an argument here? No need to inject storage then, and you're not really. You won't need the ensureEncryptionKeyIsSet() method which is a racing anti-pattern anyway. I could now call create/validate token with multiple goroutines and the state would be modified in an upredictable way.